src/ConstantInterval.cpp

```diff
@@ -0,0 +1,660 @@
+    // overflows.
+    auto consider_case = [&](int64_t a, int64_t b) {
+        int64_t c;
+        if (mul_with_overflow(64, a, b, &c)) {
+            result.min = std::min(result.min, c);
+            result.max = std::max(result.max, c);
+        } else if ((a > 0) == (b > 0)) {
+            result.max_defined = false;
+        } else {
+            result.min_defined = false;
+        }
+    };
+
+    result.min_defined = result.max_defined = true;
+    result.min = INT64_MAX;
+    result.max = INT64_MIN;
+    if (a.min_defined && b.min_defined) {
+        consider_case(a.min, b.min);
+    }
+    if (a.min_defined && b.max_defined) {
+        consider_case(a.min, b.max);
+    }
+    if (a.max_defined && b.min_defined) {
+        consider_case(a.max, b.min);
+    }
+    if (a.max_defined && b.max_defined) {
+        consider_case(a.max, b.max);
+    }
+
+    bool a_bounded_negative = a.min_defined && a <= 0;
+    bool a_bounded_positive = a.max_defined && a >= 0;
+    bool b_bounded_negative = b.min_defined && b <= 0;
+    bool b_bounded_positive = b.max_defined && b >= 0;
+
+    if (result.min_defined) {
+        result.min_defined =
+            ((a.is_bounded() && b.is_bounded()) ||
+             (a >= 0 && b >= 0) ||
+             (a <= 0 && b <= 0) ||
+             (a.min_defined && b_bounded_positive) ||
+             (b.min_defined && a_bounded_positive) ||
+             (a.max_defined && b_bounded_negative) ||
+             (b.max_defined && a_bounded_negative));
+    }
+
+    if (result.max_defined) {
+        result.max_defined =
+            ((a.is_bounded() && b.is_bounded()) ||
+             (a >= 0 && b <= 0) ||
+             (a <= 0 && b >= 0) ||
+             (a.max_defined && b_bounded_positive) ||
+             (b.max_defined && a_bounded_positive) ||
+             (a.min_defined && b_bounded_negative) ||
+             (b.min_defined && a_bounded_negative));
+    }
+
+    if (!result.min_defined) {
+        result.min = 0;
+    }
+
+    if (!result.max_defined) {
+        result.max = 0;
+    }
+
+    // Check the class invariant as a sanity check.
+    internal_assert(!result.is_bounded() || (result.min <= result.max));
+
+    return result;
+}
+
+ConstantInterval operator%(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+
+    // Maybe the mod won't actually do anything
+    if (a >= 0 && a < abs(b)) {
+        return a;
+    }
+
+    // The result is at least zero.
+    result.min_defined = true;
+    result.min = 0;
+
+    // Mod by produces a result between 0
+    // and max(0, abs(modulus) - 1). However, if b is unbounded in
+    // either direction, abs(modulus) could be arbitrarily
+    // large.
+    if (b.is_bounded()) {
+        result.max_defined = true;
+        result.max = 0;                                 // When b == 0
+        result.max = std::max(result.max, b.max - 1);   // When b > 0
+        result.max = std::max(result.max, -1 - b.min);  // When b < 0
```

Comment:
What happens if these computations overflow? i.e. if b is a single point and it's INT64MIN?